### PR TITLE
Documentation: add asset generation using dependency graph doc

### PR DESCRIPTION
**Pull request by abhinavdahiya**

### CODE REVIEW

File: Documentation/design/assetgeneration.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
+# Installer asset generation based on dependency resolution
+
+## Goal
+
+Define generation of assets of varied types in installer based on a dependency graph.
+
+## Overview
+
+The installer generates assets based on the dependency [graph](./dependency.md). To generate an asset, all its dependencies have to be generated. The installer needs to define how assets declare this dependency and how it resolves the graph when asked to generated specific targets. Targets are chosen assets that need to be generated, the installer generates all assets that the target depends on and writes to disk only the targets, consuming any inputs from disk or state from a previous run.
```
[the remark this answers is not in the document]
Comment (abhinavdahiya) [reply Documentation/design/assetgeneration.md]: >  And this paragraph is sounding a lot like Make, Bazel, Terraform, etc. ;)

mini version of these :)

File: Documentation/design/assetgeneration.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+Each asset is individually responsible for declaring its dependencies. Each asset is also responsible resolving conflicts when combining its input from disk and its state from a previous run. The installer ensures all the dependencies for an asset is generated and provides the asset with latest state to generate its own output.
+
+## State
+
+State is an internal state of the installer. It stores the contents of all the assets that have been generated. The asset is an map of all the assets that have been generated and their contents.
```
[the remark this answers is not in the document]
Comment (abhinavdahiya) [reply Documentation/design/assetgeneration.md]: :+1:

File: Documentation/design/assetgeneration.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+An example of the State:
+
+```go
+// State is map of assets
+type State struct {
+    // Verify integrity
+    Digest string
```
[the remark this answers is not in the document]
Comment (abhinavdahiya) [reply Documentation/design/assetgeneration.md]: :+1:

File: Documentation/design/assetgeneration.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        }
+    }
+}
+```
+
+## Asset
+
+An asset is the generic representation of work-item for installer that needs to be generated. Each asset defines all the other assets that are required for it to generate itself as dependencies.
+
+The asset would usually follows these steps to generate its output:
```
Comment (yifan-gu) [inline Documentation/design/assetgeneration.md]: s/follows/follow

File: Documentation/design/assetgeneration.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+1. Load components from disk. (If found, mark them from deletion, we consume our inputs.)
+
+2. Load its components from State object given by installer.
+
+3. If both the sources exist,
+
+    * Both are same: nothing to do use the State object.
+
+    * Both differ: maybe try to merge and mark itself **dirty** or error out.
```
Comment (yifan-gu) [inline Documentation/design/assetgeneration.md]: I thought we decided that we gonna just throw our hands?

File: Documentation/design/assetgeneration.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+```go
+// State is map of assets
+type State struct {
+    Root string // hash for the root asset
+
+    Objects map[string]struct{
+        Parents []string // hashes for our parents
+
+        Objects map[string] {
```
Comment (yifan-gu) [inline Documentation/design/assetgeneration.md]: Shouldn't the parent include everything but the final targets ? 
I would expect some tree-list structure, besides how is this struct related to the `Assets` struct below?

I was thinking something like:
```go
type State struct {
    Assets []Asset; // Final leaf assets
}
type Asset struct {
    Name string
    Parents []Asset
    Dirty bool
    Hash string
    Data []byte
    ...
}
```

Comment (abhinavdahiya) [reply Documentation/design/assetgeneration.md]: The state object is an hashmap containing all the assets, not just the final. Each asset will use this object to fetch its dependencies.

File: Documentation/design/assetgeneration.md (hunk 5 above)
Comment (abhinavdahiya) [reply Documentation/design/assetgeneration.md]: That's still an valid outcome.

**Comment by yifan-gu** [comment]
Discussed with @abhinavdahiya , agree on the definition of the struct in this doc, just waiting on some fix of the naming.

File: Documentation/design/assetgeneration.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+// State is map of assets
+type State struct {
+    Root string // hash for state object.
+
+    Objects map[string]AssetState
+}
+
+// AssetState is the state of an Asset.
+type AssetState struct {
+    Parents maps[string]string // hashes for our each parent.
```
Comment (wking) [inline Documentation/design/assetgeneration.md]: Why did this end up being a map?  I think it should be a slice of hash strings; see [here][1] and [here][2].

[1]: https://github.com/openshift/installer/pull/105#discussion_r207412024
[2]: https://github.com/openshift/installer/pull/105#discussion_r207695151
